### ISSUE TITLE
HPO 目標新增 pooled_average_precision 與 macro_per_item_average_precision（精確算法，吃整組抽樣的權重）

### ISSUE THREAD

**Issue by curtis-lu** [curtis-lu/recsys-demo#430]
## Parent

#426（spec）。本票實作它的**決定 G**。

## What to build

`training.hpo_objective` 多兩個合法值，讓 HPO 直接優化「把每一列當成二元預測」的 average precision：

- **`pooled_average_precision`**：val 的全部列倒在一起算一個 average precision。
- **`macro_per_item_average_precision`**：每個 item 各算一個、再平均，讓曝光量大的 item 不會蓋過曝光量小的。

預設值維持 `macro_per_item_map`，既有部署的模型選擇不變。

為什麼需要：以曝光紀錄訓練的部署，使用者主要看的是 precision–recall 面積（ADR-0024 背景），但今天 HPO 只能優化組內排序指標；query group 很小時，組內指標算不出有鑑別力的數字。

要做的事：

1. 合法值清單、HPO 評分函式、指標實作三處同步加上這兩個目標。
2. 兩者都用 scikit-learn 的 `average_precision_score`（已在釘版本的依賴裡，框架程式碼目前還沒用到它），把整組抽樣那張票產生的權重欄當 `sample_weight`。HPO 只算 val、資料在 driver 上，精確排序便宜。
3. `macro_per_item_average_precision`：**val 裡沒有正例的 item 不進平均**，由呼叫端先濾掉、不送進 `average_precision_score`。這與既有 `macro_per_item_map` 的實際行為一致（它只從正例列數 item，沒有正例的 item 根本不在分母裡）。不濾的話，`average_precision_score` 對全是 0 的 `y_true` 回 `-0.0` 並發 `UserWarning`（2026-09-20 在 scikit-learn 1.5.0 上實測），那些 item 會把平均往 0 拉。
4. HPO 的 log 印出「進入平均的 item 數／全部 item 數」。理由：item 上千、多數 item 在 val 沒有正例時，平均實際上只涵蓋少數有正例的 item——曝光少的 item 不是被保護，而是直接不在分母裡。使用者要看得到這件事。
5. 新不變量（設定層，加在一致性模組）：選了這兩個目標之一時，`dataset.val_zero_positive_group_ratio` 必須 > 0，否則在 CLI 入口擋下，訊息說明「val 在 dataset 階段已經丟掉沒有正例的 query group，average precision 會算在篩過的母體上」。
6. 文件：
   - 這兩個目標與評估報表的 `pr_auc`（#381）**名字不同、定義不同、不可對帳**：這邊是精確算法、可以跟 scikit-learn 對帳；`pr_auc` 是全量 Spark 上的分箱近似（ADR-0024 因成本選的）。刻意不叫 `pr_auc`，也不單獨叫 `average_precision`（repo 裡 AP 已經是組內排序指標的名字）。
   - `pooled_average_precision` 的已知性質：把所有列倒在一起排，會獎勵「認出哪些 entity 本來就容易有正例」，這對組內排序沒有幫助。它是這類部署主要看的指標，所以是正式目標；選它的人要知道這件事。
   - 權重是設計權重、不是無偏估計（見整組抽樣那張票）；r 小、組數少時，trial 之間的差距可能小於估計的抖動。
   - early stopping 用哪個 `metric` 照舊由使用者在 LightGBM 參數裡設，與 HPO 目標無關。

## Acceptance criteria

- [ ] 預設目標不變；沒選新目標時 `model_version` 與 main 相同
- [ ] 兩個新目標通過 CLI 入口的合法值檢查；打錯名字照舊被擋
- [ ] `pooled_average_precision` 的值與直接呼叫 `average_precision_score(y, score, sample_weight=w)` 逐值相同
- [ ] `macro_per_item_average_precision`：沒有正例的 item 不進平均、不觸發 `UserWarning`；全部 item 都沒有正例時給出明確的錯誤或定義好的值（開工時定，寫進文件），不是靜默回 `nan`／`-0.0`
- [ ] HPO 的 log 印出進入平均的 item 數與全部 item 數
- [ ] 選了新目標而 `val_zero_positive_group_ratio` 為 0（或沒寫）→ CLI 入口擋下，訊息說明原因
- [ ] `examples/ad` 以 `pooled_average_precision` 跑完一輪 HPO（小的 `n_trials`）
- [ ] 文件：fresh 讀者答得出「HPO 的 average precision 為什麼跟報表上的 `pr_auc` 對不起來」
- [ ] mutation：把 `sample_weight` 拿掉，測試紅在「與 scikit-learn 逐值相同」那條斷言

## Blocked by

- #429（整組抽樣：`val_zero_positive_group_ratio` 與權重欄）

## Out of scope

- **評估報表的 `pr_auc` 改成精確算法。** 因為 ADR-0024 已就全量排序的成本做過取捨。
- **拿 average precision 當 early stopping 的指標。** 因為 early stopping 的 `metric` 是使用者自己設的 LightGBM 參數，LightGBM 有內建的選項，不需要框架介入。
- **校準。** 因為 average precision 只吃分數的大小順序；框架不提供校準（#411）。

## 每張票都要做的

- 相容性驗證由便宜到貴：`git diff main..HEAD -- conf/base/` → 版本測試 → 示例 e2e digest
- mutation：確認測試紅在**目標斷言**上，不是紅在 fixture 建置
- 審查留一個視角不吃驗收條件：「假設 #426 或新 ADR 本身寫錯了，錯在哪」
- 全程在 `.worktrees/<name>` ＋ `feat/` branch

## 補充（2026-09-21）——新的 macro 目標讀不讀 `evaluation.metric` 的三個旋鈕

已查證的現況：`evaluation.metric` 有三個調整「以 item 為單位取平均」的旋鈕——`min_positives`（正例太少的 item 不進平均）、`weight_alpha`（依正例數加權）、`shrinkage_k`（向整體收縮）。**評估報表與診斷會讀它們，HPO 不讀**：HPO 評分函式呼叫 `compute_macro_per_item_map` 時沒有傳這三個參數，一律用預設值（0／0／0），也就是每個有正例的 item 等權。

這件事對本票的意義：上面第 4 點說「item 上千、多數 item 在 val 沒有正例時，平均只涵蓋少數有正例的 item」；再往下一步，**有正例的 item 裡，只有一兩個正例的也跟大 item 等權**，它們的 average precision 很抖，HPO 會被它們拉著走。以展示紀錄訓練的部署（item 多、多數很少被展示）最容易遇到。

本票的做法（預設，開工時不必再問）：**`macro_per_item_average_precision` 比照既有的 HPO 行為——不讀這三個旋鈕。** 理由是讓兩個 macro 目標在 HPO 裡行為一致；只讓新的那個讀，會出現「同一份設定下，HPO 的兩個 macro 目標一個吃 `min_positives`、一個不吃」的狀態。第 4 點要印的 log 多印一個數：進入平均的 item 裡，正例數的最小值與中位數，讓使用者看得到平均有多少是建立在一兩個正例上。文件寫明：遇到這種情況改用 `pooled_average_precision`（或既有的 `mean_ap`）比較穩。

**不在本票範圍**：讓 HPO 的 macro 目標（新舊兩個一起）讀 `evaluation.metric` 的旋鈕。因為那會改變既有部署在設了這些旋鈕時的 HPO 結果（`model_version` 不變、選出來的 trial 卻可能不同），要不要做、怎麼不靜默地做，是另一個決定。

**Comment by curtis-lu** [comment]
## 開工前釐清（2026-09-22）——計算成本與四個定案

開工前確認了兩件事：票上的前提在 #429 merge 之後還對不對，以及兩個新目標在 HPO 裡會不會很貴。

### 前提都還成立

- 權重欄是 `zero_positive_group_weight`（`core/consistency.py::ZERO_POSITIVE_GROUP_WEIGHT_COL`），有正例的組是 1，留下來的無正例組是 1／r。只有 val 的 r > 0 時，val 表才有這一欄；定案 4 的閘保證選了新目標時 r > 0。
- `CONTEXT.md` 已經有 `pooled_average_precision`／`macro_per_item_average_precision` 的詞條，不用再加。
- 評估端已經有同一類的檢查：A46（test 的 r ＝ 0 時擋 `prediction_quality`）。本票的閘擋的是同一件事（算在已經篩過的母體上）。但 A46 掛在 evaluation 指令上，本票的閘不是，位置見定案 4。
- `hpo_objective` 會算進 `search_id`（`core/versioning.py:252`），換目標就開新的 study，不會接到舊目標的進度。
- `diagnosis/hpo/` 只把目標名稱原樣記下來，不用改。

### 算 average precision 本身不貴

本機量的（8 核、load ≈ 2、scikit-learn 1.5.0、亂數資料、每種情況量一次）：

| val 列數 | item 數 | `pooled` | `macro` 先按 item 排序一次 | `macro` 每個 item 各掃一遍全部的列 |
|---|---|---|---|---|
| 500 萬 | 22 | 1.5 s | 1.3 s | 1.3 s |
| 500 萬 | 1000 | 1.5 s | 1.4 s | **5.4 s** |
| 2,000 萬 | 22 | 10.2 s | 5.7 s | 5.6 s |
| 2,000 萬 | 1000 | 10.0 s | 5.9 s | （沒量） |

比較基準：生產一個 trial 光訓練就要約 7 分鐘（`docs/notes/2026-09-07-training-pipeline-profiling.md` §4.3，20 個 trial 外推 2.43 小時）。所以每個 trial 算 AP 多花約 2–5%。一個 trial 只算選中的那一個目標，不會兩個都算。

**會變貴的是 val 的大小。** 新目標要求 val 的 r > 0，val 就得留下沒有正例的組。以組數算（各組大小差不多時，列數也放大同樣倍數）：

```
p ＝ val 裡有正例的組佔幾成
val 放大成原本的 (p + r × (1 − p)) ／ p 倍
例：p ＝ 10%、r ＝ 0.5 → 5.5 倍
```

每個 trial 都要把整個 val 矩陣走一遍去預測。生產的 val 矩陣估算是 37–89 GiB（`io/disk_matrix.py`，driver 128 GB）。這個數字是在 #429 之前估的，當時 val 只留有正例的組，也就是現在的預設 r ＝ 0。放大 5 倍就放不進頁快取（作業系統拿記憶體暫存磁碟內容的地方），每個 trial 都要從磁碟重讀。`data/_scratch` 空間不夠時，`tune_hyperparameters` 本來就會 raise（`docs/pipelines/training.md` §2 第 5 點），但頁快取不夠不會報錯，只會變慢。

### 四個定案

1. **成本只寫文件、加 log，不加限制。** 文件（`docs/pipelines/training.md` 的 HPO 目標那段）寫上面的放大倍數，並說明 r 就是成本與雜訊之間的取捨：r 越小，val 越小，但權重 1／r 越大，分數越抖。HPO 開始前的 log 多印一行：val 裡「有正例的組共幾列」與「留下的無正例組共幾列」。**判斷一組有沒有正例要看 label，不能看權重**：r ＝ 1 時兩種組的權重都是 1。不做「val 太大就警告／擋下」，因為門檻要看機器，框架訂不出來。
2. **val 完全沒有正例 → 搜尋開始前 raise，只針對兩個新目標。** 這就是驗收條件裡「開工時定」的那一項。「所有 item 都沒有正例」就等於「整個 val 沒有正例」，一個檢查就同時涵蓋 `pooled` 與 `macro`。檢查放在 `study.optimize` 之前，只做一次，不放在每個 trial 裡。舊的兩個目標維持回傳 0.0，不改它們的行為。為什麼不回傳定義好的值：回傳常數的話，每個 trial 分數都一樣，TPE 學不到東西，第一個 trial 就直接勝出，而且整個過程不會報錯。
3. **`examples/ad` 永久改成 `pooled_average_precision`**（`examples/ad/conf/base/parameters_training.yaml`），並重產基準 digest，做法同 e9a2e4f9。它的 val r 已經是 0.5，會通過定案 4 的閘。舊目標 `macro_per_item_map` 仍由主線 `conf/base` 的 e2e 測到。
4. **「選了新目標但 val 的 r ＝ 0（或沒寫）」放進所有 CLI 入口都會跑的 `validate_config_consistency`**，不掛在 training 指令上。理由是 r 在 dataset 階段就生效了：如果等到 training 才擋，dataset 已經白跑一次，而且改 r 會翻 `base_dataset_version`，一樣得重跑 dataset。代號：2026-09-22 查的時候，`core/consistency.py` 最大到 A46，沒有 open PR，下一個空號是 A47。開 PR 前再查一次有沒有人先用掉。

### 實作時直接照做（不必再問）

- **`macro` 先按 item 排序一次再切段**，不要對每個 item 各做一次布林遮罩。上表 1000 個 item 時，遮罩的做法慢將近 4 倍，item 越多差越大。
- **同分怎麼處理照 scikit-learn。** `average_precision_score` 把同分的列合成一個門檻，不用 item 或 `event` 去排同分的先後。驗收條件要求跟 scikit-learn 逐值相同，所以只能這樣做。這個處理方式與列的順序無關：20 萬列、50 種分數、權重 1 與 1／0.3，打亂 5 次，最大差 2e-17。所以新目標不需要 `items`／`event_keys` 來排同分。文件要寫一句：這和 repo 裡排序指標「同分依 item 排」的規則不同。
- `pipelines/training/nodes.py:641–645` 的註解說「both HPO scores skip a group without a positive by construction」，選新目標時就不對了，要改。
- `extract_Xy_with_groups` 目前不回傳權重欄。HPO 需要把 val 的 `zero_positive_group_weight` 讀出來、跟列對齊，只在選了新目標時才讀。

**Comment by curtis-lu** [comment]
## 實作時審查找到的事（2026-09-22，PR #442）

審查特別留一個視角，不看驗收條件，只問「假設 #426 或 ADR-0025 本身寫錯了，錯在哪」。找到兩處規格層級的問題（實作照規格寫了，PR 內只改了文件），另外找到一個實作 bug，已在 PR 內修好。

以下數字都用 `examples/ad` 的實際 val 量出來（690 個請求，2,815 列）。先用快取的 `.bin` 與 `hpo_trials.json` 的參數重播 5 個 trial，重播出的 pooled 分數跟端到端 log 裡的分數逐位相同，再用這 5 個 trial 的預測來量。

### 1. `macro_per_item_average_precision` 看不到同一組裡 item 之間誰排前面（規格層級）

每個 item 只拿自己的列互相比。給某個 item 的所有分數都加上同一個常數，它的值完全不變，但每一組裡的排序（框架真正的輸出）已經整個變了。實測（trial 0，每個 item 各加一個隨機常數）：

| 指標 | 加常數前 | 加常數後 |
|---|---|---|
| `macro_per_item_average_precision` | 0.3735 | **0.3735（不變）** |
| `mean_ap` | 0.750 | 0.551 |
| `pooled_average_precision` | 0.354 | 0.159 |

所以它量的是「同一個 item 的列裡，哪些比較可能是正例」，不是「組內怎麼排」。票面與 ADR-0025 決定 4 只說它「讓曝光量大的 item 不會蓋過曝光量小的」，沒提這件事。

- **PR 內做了**：寫進 `docs/pipelines/training.md` §3.2 與函式的 docstring。
- **沒做**：ADR-0025 沒有改；這個目標也照樣提供。要不要在 ADR 加一段更正、要不要繼續提供這個目標，留給之後決定。

### 2. `examples/ad` 換目標的理由不成立

PR 原本寫的理由是「一次請求只有幾個候選，組內排序指標在這裡鑑別力低」。這在示例的資料上不成立：

- 有正例的 420 個請求，多半有 3～6 個候選（組大小：組數＝1:4、2:25、3:76、4:127、5:105、6:83）。全部候選都是正例、不管怎麼排 AP 都是 1 的，只有 4 個（1%）。
- 以請求為單位重抽 400 輪，原本的贏家仍然贏的比例：`macro_per_item_map` 0.96、`mean_ap` 0.63、`pooled_average_precision` 0.48。pooled 選中的 trial 只贏第二名 0.0013，而配對差的標準差是 0.0085，等於擲硬幣。

示例仍然用 `pooled_average_precision`，理由改成照實寫：讓端到端實跑走到這條程式路徑，不是因為它在這份資料上選模比較好。票面「query group 很小時，組內指標算不出有鑑別力的數字」這個前提本身有條件，一組只有一兩列時才成立；這份資料沒有測到那種情況。

### 3. val 的 dataset 版本可能比設定檔舊（實作 bug，已修）

A48 檢查的是設定檔，但 training 讀的是磁碟上的 dataset 版本（`latest`，或 `--base-dataset-version` 指定的版本），不會照設定檔重算版本（`__main__.py` 的 `resolve_base_dataset_version`）。兩者可能對不上：

- **那版 val 是 r ＝ 0 建的**：它沒有權重欄。原本這一欄會被靜默略過，要把整個 val 矩陣讀完（生產上是幾十 GiB）才以 `KeyError` 失敗。現在讀矩陣之前就 raise，訊息會說明原因與修法。
- **那版 val 是用別的 r 建的**：權重照建表當時實際的 1／r 算，所以分數是對的，但 log 原本印的是設定檔的 r。現在同時印出設定檔的 r 與 val 裡實際的權重，兩個數對不上時一眼就看得到。

**Comment by curtis-lu** [comment]
## 更正與定案：`macro_per_item_average_precision` 保留（PR #443，2026-09-22）

前面留言說它「看不到同一組裡 item 之間誰排前面」，還把「要不要繼續提供」列為未決，那是拿**組內排序**的尺去量一個**二元預測**的目標。準確的說法是：它判斷一列的分數算高還是算低，只拿它跟同一個 item 的其他列比，所以不檢查不同 item 之間的分數比不比得起來。

**已定案：兩個目標都保留，分工如下：**
- 分數要跨 item 比較（同一次請求挑分數最高的 item，或全部 item 共用一條門檻）：選 `pooled_average_precision`。
- 每個 item 的分數各自使用（例如每個 item 各自定門檻）：選 `macro_per_item_average_precision`。

說法以 `docs/pipelines/training.md` §3.2（附數值例子）與 ADR-0025 決定 4 的補記為準。
